Add ConstraintVerifier infomation to OptaPlanner documentation (#15611)
quarkusio/quarkus: docs/src/main/asciidoc/optaplanner.adoc

This approach ensures that the test runs long enough to find a feasible solution, even on slow machines.
But it does not run a millisecond longer than it strictly must, even on fast machines.

> 

Reviewer: This new content should have a separate section title than the TimeTableResource part and it should come first because it's more important to teach users to test their constraints:

Suggestion: Under the section title "Test the application", create 2 subsections:
- "Test the constraints" => the new content
- "Test the solver" => the original content

> Additionally, if you use a `ConstraintProvider` and have `org.optaplanner:optaplanner-test` as a dependency in your pom, you can test per constraint using a
> `ConstraintVerifier`.
> 
> Add the `org.optaplanner:optaplanner-test` dependency to your `pom.xml` dependencies section:

Reviewer: Style, use "active tense" to avoid "you" whenever possible (even though "you" is always better than "we" or "I").

Start with the benefit for the user (test per constraint), before descibing the conditions (add optaplanner-test).

In the context of this text, the use of ConstraintProvider is a given, so no need to repeat it as it makes it more difficult.

```suggestion
To test each constraint in isolation, add `org.optaplanner:optaplanner-test` as a dependency in the `pom.xml`:
```

@QuarkusTest
class TimeTableConstraintProviderTest {

    private static final Room ROOM = new Room("Room1");
    private static final Timeslot TIMESLOT1 = new Timeslot(DayOfWeek.MONDAY, LocalTime.of(9,0), LocalTime.NOON);
    private static final Timeslot TIMESLOT2 = new Timeslot(DayOfWeek.TUESDAY, LocalTime.of(9,0), LocalTime.NOON);

    @Inject
    ConstraintVerifier<TimeTableConstraintProvider, TimeTable> constraintVerifier;

    @Test
    void roomConflict() {
        Lesson firstLesson = new Lesson(1, "Subject1", "Teacher1", "Group1");
        Lesson conflictingLesson = new Lesson(2, "Subject2", "Teacher2", "Group2");
        Lesson nonConflictingLesson = new Lesson(3, "Subject3", "Teacher3", "Group3");

        firstLesson.setRoom(ROOM);
        firstLesson.setTimeslot(TIMESLOT1);

        conflictingLesson.setRoom(ROOM);
        conflictingLesson.setTimeslot(TIMESLOT1);

        nonConflictingLesson.setRoom(ROOM);
        nonConflictingLesson.setTimeslot(TIMESLOT2);

        constraintVerifier.verifyThat(TimeTableConstraintProvider::roomConflict)
                .given(firstLesson, conflictingLesson, nonConflictingLesson)
                .penalizesBy(1);
    }

}
----

> This test verifies that the constraint `TimeTableConstraintProvider::roomConflict`, when given three lessons in the same room, where two lessons have the same timeslot, it penalizes the score by 1.

Reviewer: This should point out that it tests the matchWeight and ignores the constraintWeight. In fact "it penalizes the score by 1" is wrong.

```suggestion
This test verifies that the constraint `TimeTableConstraintProvider::roomConflict`, when given three lessons in the same room, where two lessons have the same timeslot, it penalizes with a  match weight of `1`. So with a constraint weight of `10hard` it would reduce the score by `-10hard`.

Notice how `ConstraintVerifier` ignores the constraint weight during testing - even if those constraint weights are hard coded in the `ConstraintProvider` - because constraints weights change regularly before going into production. This way, constraint weight tweaking does not break the unit tests.
```
